### ISSUE TITLE
Add a new origamiCategory that allows css from other components, and templates!

### ISSUE THREAD

**Issue by chee** [Financial-Times/origami#45]
## The idea

Right now we have 4 origamiCategory values: `primitives`, `layouts`, `components` and `utilities`.

I'm proposing we add a new one! I don't have a good name for it yet but for now i'm going to use `composites`, meaning "made out of parts" (because these types of components generally build on other origami components).

Because it's completely new to the spec, we can relax the rules around templates and css-output just for that category without breaking backwards compatibility. I think we should relax the following rules in the following ways:

### 1.

In the [markup spec](https://origami.ft.com/spec/v1/markup/#templates) we say this: 

> "Origami components **should not** provide templates directly, outside of HTML within component demos and the readme. But where they do, templates **must** be in Mustache format.".

I think we should say that components with the category of `composites` **may** use templates! We shouldn't be more specific than that for now.

The current Templates section shouldn't apply to `composites`. `composites` should be allowed to choose whatever is right for them, as long as they provide demos and markup in their README. We should note that changes to a template must be considered breaking changes.

### 2.

Basically every rule about how components can use, override or mention "other components" in the [sass spec](https://origami.ft.com/spec/v1/sass/) should not apply to `composites`

### um

I couldn't find where we say they can't output the CSS of another component, but let's turn that off for `composites` too!

## Work involved

1. Create a PR updating the spec to allow these things to exist
2. Communicate with our customers who build non-origami components of this kind and see if they need changes to the PR or what we can do to help
3. maybe try converting an n-component?
4. merge it in???

**Comment by notlee** [comment]
>I couldn't find where we say they can't output the CSS of another component, but let's turn that off for composites too!

Yeah I thought their was something specific but couldn't find it! I suppose it's covered by:
>[CSS selectors] must also be prefixed with the component name and written as hyphen separated, lowercase strings
If it's from another component it's not prefixed with the component name

**Comment by notlee** [comment]
Will this need a change to occ to publish templates?

**Comment by chee** [comment]
i had a discussion with customer products today about an approach they could take with n-compoents, which would interfere with one of the aims of this proposal.

at the moment a lot of their components depend on [`n-ui-foundations`](https://github.com/Financial-Times/n-ui-foundations), and depend on things that depend on `n-ui-foundations` in order to share styles. 

we discussed a new approach where the n-components would be npm-first, and list the origami components as `peerDependencies`, and require the app to provide the correct origami versions. this, combined with dependabot to keep the origami versions up to date in the apps, would invert the depenency tree ridding them of n-ui-foundations while hopefully retaining the page-kit shared-style strategy caching benefits.

but it would mean that these n-components could no longer be origami components, because we require origami components to be bower-first.

it's a big bit of work on their part, but i think it might be a good idea. if they choose that path, it probably means we should revisit this proposal.

**Comment by notlee** [comment]
Would you mind expanding? I'm not sure I understand how using `peerDependencies` removes the need for `n-ui-foundations`. Would any styles in `n-ui-foundations` be broken up into `n-things`? How would `peerDependencies` on `n-things` differ from using a caret range? Is it about enforcing a flat dependency tree?

**Comment by chee** [comment]
yes, it's about making it easier to have a flat dependency tree once they switch to npm.

it maybe doesn't rid them of n-ui-foundations (it also provides some config for the o-components), but maybe it eases the stranglehold `n-ui-foundations` has on the dependency tree of ft apps

`n-ui-foundations` [depends on](https://github.com/Financial-Times/n-ui-foundations/blob/master/bower.json#L13-L23) shared origami components, and n-components and apps depend on n-ui-foundations.

![a digraph showing n-ui-foundations dependency tree](https://user-images.githubusercontent.com/178266/84022540-f979d380-a975-11ea-8a53-d9f2d563d2b4.png)

it is, among other things, a way of ensuring all the versions of some common origami components are the same across all the apps so that the page-kit caching strategy can end up with the same artifact.

npm wants a flat tree, but doesn't demand it. it uses a feature of node js's module resolution system to do the trick. in nodejs, if you say require("monkey") it will look in:
- `./node_modules/monkey`
- `../node_modules/monkey`
- `../../node_modules/monkey`
- `../../../node_modules/monkey`

 all the way up until it can't go any further because it's at `/node_modules/monkey`
so by default npm installs things to `./node_modules/package_name` but if it hits a conflict, rather than FReAKiNG OuT it puts the conflicting version required in a `node_modules` subfolder of the thing that depends on it!

for example: 

app depends on monkey@2.0.0 and banana@1.0.0
monkey@2.0.0 depends on banana@2.0.0
npm creates this tree:

```
app
 |  - node_modules
    | - monkey (version 2)
      | - node_modules
        | - banana (version 2)
    | - banana (version 1)
```

so now because of the nodejs module resolution system when app requires banana, it gets version 1. but when monkey requires banana, it gets version 2

`peerDependencies` will ensure an origami component is only installed once. the app will have to install them, and if there are conflicts then the user will be unable to install the packages required to satisfy them. a [future version](https://github.com/npm/rfcs/blob/7650ec4646481f0242d1081601d543d1c5e88c7f/accepted/0000-install-peer-deps.md) of npm will install these automatically and fail if there is a conflict.

so there will only be one, rather than nested ones.

dependabot will file pull requests to keep them up to date in the apps, and the apps and the components will be depending on the origami components they actually use rather than depending on an intermediary.

but this is an unexplored idea. it's not something that will involve origami, it's a customer decision. but it's something that they might look into, and it would render this proposal unhelpful.

so what i'm saying is, let's bare in mind that this proposal makes some assumptions about what would benefit our customers and we should pause it until we have some more buy in from other teams.

**Comment by notlee** [comment]
>dependabot will file pull requests to keep them up to date in the apps, and the apps and the components will be depending on the origami components they actually use rather than depending on an intermediary.

cool, definitely 👍 

Do you know if customer products will be writing up a proposal to investigate this approach? I'm wondering for example if it would ensure flat dependencies where origami components depend on other origami components that aren't a peer dependency of an n-thing

**Comment by notlee** [comment]
oh, and thanks for the super detailed explanation :)

**Comment by JakeChampion** [comment]
Problems:

1. Origami via npm currently can have an issue where it's dependency tree is flat, we solve this with custom tooling but would be better to rely on an npm in-built feature for solving this if possible.

2. Component composition - combining the markup and css classes of multiple components into a new component is not compliant with the origami specification and the reason it is not specification compliant is... (need explanation here) - We currently solve this for smaller component such as o-buttons by using "content mixins". This is an issue because we are wanting to make it easier for Customer Products to make more Origami components. If customer products are not wanting to make more Origami components, then we think this is not a problem worth solving.

**Comment by chee** [comment]
marking this as blocked, because i think we'd like to explore switching to npm-first and stop components from depending directly on one another (and instead using peerDependencies).

**Comment by chee** [comment]
let's v2 origami instead
